### DISCUSSION TITLE
Rendering into FBO using Shared Context

### DISCUSSION THREAD

**Discussion by roxlu** [google/filament#5930] [General]
I've been integrating Filament into an application where I want to use Filament to do it's awesomeness regarding 3D rendering and use my current engine for some other rendering, video playback, 2D, layouting, etc. 

If I'm correct this is possible. With the abilty to use a shared (OpenGL) context it looks like one can "share" Filament with a another context and share e.g. textures, fbos, etc.  
But as @romainguy noted, the internal state is and should not be exposed to be able to transparently support multiple backends (#1299). 

For some use cases being able to use the output of Filament seems like a obvious feature. Maybe, one solution that might work is to render a view into a FBO (which can be shared via a shared context) and then use the resulting FBO texture with my current engine.  Please correct me if I'm wrong.

To make this work, I would need to get the texture ID into which the view is rendered. After searching through the issues I found some posts that are related:

[Filament + GLFW: how to share context and managing swapchain (my  post), 1894](https://github.com/google/filament/issues/1894)
[Get OpenGL texture id from Texture instance, 1784](https://github.com/google/filament/issues/1784)
[Access underlying buffer object associated with a texture, 1299](https://github.com/google/filament/issues/1299)

Issue #1299 contains a lot of information but it's been a while that there was any activity.  The suggested solution is to create a `Stream` with an external texture id. Is this still the best solution there is, also when dealing with a Framebuffer? 

Thanks

**Comment by romainguy** [comment]
We don't want to expose the underlying GL objects because the engine needs to work with other graphics APIs. Which platform are you targeting? On Android you can have Filament render into a `Surface` backed by a `SurfaceTexture` which you could then use in your OpenGL engine. Other platforms have similar solutions but I'm not familiar with them (there's always PBOs I suppose). I'd rather you tried this approach instead.

**Comment by roxlu** [comment]
I'm targeting OpenGL, Desktop; I'm in full control of the system where I'm running this application. I'm not sure how PBO fits in here as I've mainly used those when using asynchronous cpu->gpu transfers. Is there a similar approach like SurfaceTexture for OpenGL ?

**Comment by roxlu** [comment]
I’ve been searching existing issues and the source code but although I found a couple of issues where @romainguy mentions the use of an external texture and surface I’m not certain which classes/types I should look into. Especially  the use of a Surface in relation to my use case (glfw, existing shared GL context) is unclear to me.

I read that using an external texture via a Stream triggers a copy? Not sure if this is the case but I suspect that to have a performance penalty.

Can you share any suggestions what part of the API I should dive into?

**UPDATE**
When you see this comment in Stream.h it seems that a Stream is indeed used to copy data. The comments says that data is sampled from the external texture. Therefore this is not what I need as I need Filament to write into my texture (fbo color attachment).

```c++
 /**
 * Creates a TEXTURE_ID stream. This will sample data from the supplied
 * external texture and copy it into an internal private texture.
 *
 * @param externalTextureId An opaque texture id (typically a GLuint created with glGenTextures)
 *                          In a context shared with filament. In that case this texture's
 *                          target must be GL_TEXTURE_EXTERNAL_OES.
 *
 * @return This Builder, for chaining calls.
 *
 * @see Texture::setExternalStream()
 */
Builder& stream(intptr_t externalTextureId) noexcept;
```

**Comment by roxlu** [comment]
I would love to discuss this topic further but if I got it right by reading some issues other people posted, I think there is no intention at the moment, to provide a way that gives users the ability to access driver specific handles/IDs? 

I think, even though the design and goal is to create an engine that hides the internals,  in some cases it makes sense to allow users access the underlying IDs/handles. 

I went ahead and added a function that returns the ID of a texture. Now, I'm able to use Filament to render into a FBO and use the result in my own engine. I've created a video (no more flickering ^.^): https://imgur.com/a/BAqpGgQ

**Comment by roxlu** [comment]
I created a fork and added a function that allows you to retrieve a handle to a texture ID.  I'm not aware of a lot of the design choices that were made regarding the driver and backend system and I'm sure there would have been a different way to implement this feature. I would like to discuss this even though it will not be part of Filament. 

As I'm sure this might be interesting for others who are trying to use Filament in their engine, this is the feature branch that provides a function to a texture handle:

https://github.com/roxlu/filament/tree/rox/feature-get-texture-id

**Comment by roxlu** [comment]
Just stumbled upon a great read that shares some good info on multi threaded design and use of OpenGL contexts: https://gstreamer.freedesktop.org/documentation/additional/design/opengl.html?gi-language=c

**Comment by roxlu** [comment]
See [this update](https://github.com/google/filament/issues/1921#issuecomment-638165840) which explains a bit more how to create and use shared contexts.

**Comment by pixelflinger** [comment]
At this point this looks like more a discussion, I don't think there is any action items for us.
